### PR TITLE
feat: add Accordion and Collapsible for web (pre-release)

### PR DESCRIPTION
**Pull request by divyanshu013**
> **Note**
>
> The story is marked internal

- Adds support for `CollapsibleLink` with animations
- Fixes a slight jank in animation
- More PRs to follow

https://github.com/razorpay/blade/assets/6682655/5e2f8562-ee34-47ab-9e2d-e4254610fd24

### CODE REVIEW

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  isExpanded: boolean;
+  direction: CollapsibleProps['direction'];
+};
+
+// TODO: refactor common transition properties
+const StyledCollapsibleLinkIcon = styled(BaseBox)<StyledCollapsibleLinkIconProps>((props) => {
+  const { isExpanded, direction, theme } = props;
+  let transformExpanded, transformCollapsed;
+  if (direction === 'bottom') {
+    transformExpanded = 'rotate(0.5turn)';
```
Comment (anuraghazra) [inline packages/blade/src/components/Collapsible/CollapsibleLink.tsx:28]: TIL `turn` is a valid unit.

File: packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx [hunk 2]
```diff
@@ -61,17 +62,38 @@ const meta: Meta<CollapsibleProps> = {
...
 
-const CollapsibleTemplate: ComponentStory<typeof CollapsibleComponent> = ({ ...args }) => {
+const CollapsibleButtonTemplate: ComponentStory<typeof CollapsibleComponent> = ({ ...args }) => {
   return (
     <CollapsibleComponent {...args}>
-      <CollapsibleButton>Answer to life, universe and everything</CollapsibleButton>
+      <CollapsibleButton>Greatest song ever</CollapsibleButton>
+      <CollapsibleBody>
+        <Text>And as we wind on down the road</Text>
+        <Text>Our shadows taller than our soul</Text>
```
Comment (anuraghazra) [inline packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx:71]: Missed opportunity to rickroll blade consumers.

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx (hunk 1 above)
Comment (anuraghazra) [reply packages/blade/src/components/Collapsible/CollapsibleLink.tsx:28]: Is browser support okay tho? maybe lets fallback to deg only.

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    transformCollapsed = undefined;
+  } else {
+    transformExpanded = undefined;
+    transformCollapsed = 'rotate(0.5turn)';
+  }
+
+  return {
+    display: 'flex',
+    alignItems: 'center',
+    transform: isExpanded ? transformExpanded : transformCollapsed,
```
Comment (anuraghazra) [inline packages/blade/src/components/Collapsible/CollapsibleLink.tsx:38]: 🤔 why do we need two variables? can't we do isExpanded 90deg or 0deg?

**Review by anuraghazra — APPROVED** [review]
Looks good, saw few TODO comments tho make sure to resolve those,

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx (hunk 1 above)
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/CollapsibleLink.tsx:28]: Don't see turn specific compatibility on rotate, I think it should be widely supported as deg.

File: packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx (hunk 2 above)
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx:71]: Haha, will keep in mind for future 😸

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx (hunk 3 above)
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/CollapsibleLink.tsx:38]: We need to derive these based on the `direction` prop so kept two variables which derive these based on `direction` and then here I set them based on `isExpanded`

**Comment by divyanshu013** [comment]
Yup, planned TODO items in future PR(s)

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx (hunk 3 above)
Comment (kamleshchandnani) [reply packages/blade/src/components/Collapsible/CollapsibleLink.tsx:38]: still didn't understand 😅 ?

File: packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx [hunk 4]
```diff
@@ -119,7 +120,12 @@ const CollapsiblePanel = ({ children }: CollapsiblePanelProps): ReactElement =>
       defaultIsExpanded={defaultIsExpanded}
       onTransitionEnd={onTransitionEnd}
     >
-      {children}
+      <Box
+        // need margin here to prevent jank
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx:124]: what is the reason for jank?

File: packages/blade/src/components/Collapsible/CollapsibleLink.tsx (hunk 3 above)
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/CollapsibleLink.tsx:38]: The orientation of chevron icon inverts based on the direction collapsible expands in. `transformExpanded` and `transformCollapsed` therefore need to swap their corresponding expanded and collapsed values.

https://github.com/razorpay/blade/assets/6682655/f33a787c-e62d-4f7a-9a86-aa3ab8da29b0

I've also added this as a comment in #1296

File: packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx (hunk 4 above)
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx:124]: So we need a margin inside the outside wrapper so this is included in height calculations and prevents jank. Earlier I tried following for spacing which didn't work:

- `gap` on the container flex (gap isn't included in height calculation and led to jank noticed in PR1)
- `paddingBottom` / `marginBottom` on the collapse link (again it wouldn't get counted in the height calculations)

So we needed a margin on the collapsible body's inner container, that ultimately worked 😅

File: packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx [hunk 5]
```diff
@@ -61,17 +62,38 @@ const meta: Meta<CollapsibleProps> = {
...
+    </CollapsibleComponent>
+  );
+};
+
+const CollapsibleLinkTemplate: ComponentStory<typeof CollapsibleComponent> = ({ ...args }) => {
+  return (
+    <CollapsibleComponent {...args}>
+      <CollapsibleLink>Answer to life, universe and everything</CollapsibleLink>
       <CollapsibleBody>
         <Text>42</Text>
```
[the remark this answers is not in the document]
Comment (divyanshu013) [reply packages/blade/src/components/Collapsible/Collapsible.stories.internal.tsx:89]: Done, updated all stories in #1296

**Review by kamleshchandnani — DISMISSED** [review]

**Review by anuraghazra — DISMISSED** [review]

File: .changeset/rich-hounds-grow.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+---
+"@razorpay/blade": minor
+---
+
+feat: add `Accordion` and `Collapsible` pre-release
+
+- `Accordion` and `Collapsible` are available on web as pre-release
+- You may import and use the following: `Accordion_PRE_RELEASE` and `Collapsible_PRE_RELEASE`
+- The stories are marked internal for now, till the stable release, visit them from [here](https://blade.razorpay.com/?path=/story/guides-intro--page&globals=showInternalComponents:true;measureEnabled:false)
```
Comment (kamleshchandnani) [inline .changeset/rich-hounds-grow.md:9]: maybe just enable them? just make the component name as `Accordion_PRE_RELEASE`?

Comment (divyanshu013) [reply .changeset/rich-hounds-grow.md]: Done in https://github.com/razorpay/blade/pull/1277/commits/c70905cb15f43025db797c0785265e3615fbadfc

**Review by anuraghazra — DISMISSED** [review]

**Review by anuraghazra — APPROVED** [review]

**Pull request merged**